### DISCUSSION TITLE
New ideas for Riskfolio-Lib

### DISCUSSION THREAD

**Discussion by dcajasn** [dcajasn/Riskfolio-Lib#9] [Ideas]
Hi people, I would like to ask you what new features, portfolio optimization models or techniques would you like to add to Riskfolio-Lib. I think that I cover all available convex portfolio optimization models but maybe you know more models.

**Comment by RouvenG** [comment]
Thanks for this project. Do you plan to reference the original papers/references for the implemented optimization strategies? I'd love to have those for further research.

**Reply by dcajasn** [reply]
Hi @RouvenG, all papers that I used as reference for each portfolio optimization model are cited. You can see the bibliography in the documentation on the bottom of each page.

**Comment by RouvenG** [comment]
I used the examined the [Maximum Diversification Portfolio ](https://www.tobam.fr/wp-content/uploads/2014/12/TOBAM-JoPM-Maximum-Div-2008.pdf) in some research I did and couldn't find it it directly.  It is a variation of the Minimum Variance Portfolio, so it is kind of here already. I could implement it and create a pull request, if you like.

**Reply by dcajasn** [reply]
Personally, maximum diversification portfolio (MDP) doesn't have sense for me, for this reason I didn't implement it in Riskfolio-Lib. MDP is just the Sharpe optimization model where we change mean vector for standard deviation vector. If you want to use it through riskfolio-lib, after run assets_stats() method you must add the code port.mu = np.sqrt(np.diag(port.cov)).reshape(-1,1) and use obj = 'Sharpe' and the result are the MDP weights.

**Comment by DanielGoncalves47264** [comment]
Hi @dcajasn, congratulations on the great library. I was wondering if you were thinking of implementing the possibility of adding a target volatility and leverage for HRP or NRP, to fight against the low returns associated with both allocations.
Best Regards

**Reply by DavidNikolas96** [reply]
I was wondering the same thing. Would be really great to have this feature or another way to implement it 💯

**Reply by dcajasn** [reply]
Hi @DanielGoncalves47264 and @DavidNikolas96 , about your suggestions:

- About target volatility, this feature is not part of the HRP because the algo uses naive risk parity. I've never investigated that posibility for this reason and also I've never read a paper working on that feature. It could be implemented in NCO but it makes the model too complex for basic users because it's posible that in most cases the algo couldn't find a solution.
- About leverage, I saw that feature for HRP in Mlfinlab and Portfoliolab (libraries from Lopez de Prado) using negative weights, but it's not practical for real applications because you need to give in advance the assets that you are going to short. In NCO case, I think it is possible depending on the objective function but I didn't implement it because the algo could have some problems to meet the leverage constraints.

Best
Dany

**Reply by DavidNikolas96** [reply]
Thanks for your quick reply and clarifying. But by doing this overall volatility of the portfolio is really low. The possibility of getting higher returns albeit increasing risk is not neglected. What I was thinking was using leverage to gain larger exposure for example in fixed income products which tend to be less risky and leverage them up to get equity-like returns. This would in fact increase the risk of the portfolio but only slightly. On the positive site one would benefit from higher returns. Do you think there is an easy way to implement that? 
Regarding papers, there is a paper called "towards robust portfolios" from munich Re which specifically mention target volatility in HRP. However they do not specify how they implement it

**Reply by DavidNikolas96** [reply]
By using leverage every asset would still be able to equally contribute to risk (standard deviation) with the constraint that the volatility of e.g. fixed income products would be higher than before due to leverage. I cannot wrap my head around how to do it...

**Reply by Nicky9529** [reply]
Besides the "ideal" leveraged solution, dealing with HERC for example, would it be possible to implement Richard and Roncalli's Constrained Risk Budgeting algorithm to overweight/underweight clusters to target a given volatility (or other risk measure) ?

**Reply by dcajasn** [reply]
> Thanks for your quick reply and clarifying. But by doing this overall volatility of the portfolio is really low. The possibility of getting higher returns albeit increasing risk is not neglected. What I was thinking was using leverage to gain larger exposure for example in fixed income products which tend to be less risky and leverage them up to get equity-like returns. This would in fact increase the risk of the portfolio but only slightly. On the positive site one would benefit from higher returns. Do you think there is an easy way to implement that?
> Regarding papers, there is a paper called "towards robust portfolios" from munich Re which specifically mention target volatility in HRP. However they do not specify how they implement it

If they don't show how they do that in formulas or as a step in the HRP algo maybe is just an idea they have but they don't know how to do it or even if it is possible. About leverage, the HRP wasn't designed to build a leverage portfolio, to get leverage you need to short in advance some assets, this means that the model can't decide which are the best assets to short. About fixed income, this kind of products can't be introduced in a portfolio optimization directly, first you need to model them using a risk factor model and then build they covariance matrix based on the factor model.

**Reply by dcajasn** [reply]
> Besides the "ideal" leveraged solution, dealing with HERC for example, would it be possible to implement Richard and Roncalli's Constrained Risk Budgeting algorithm to overweight/underweight clusters to target a given volatility (or other risk measure) ?

Hi @Nicky9529 risk budgeting is designed to split risk contribution among assets, not for target volatility.

**Reply by jonx86** [reply]
can't you just scale up the weights using the leverage target.

**Comment by iron-marshal** [comment]
I think the only thing that is missing is an Out of Sample testing module. There are so many permutations of weights with all of the models and risk measures that you need some way of assessing out of sample performance. You could approach this via Monte Carlo or using the approach of simulated covariance matrixes (CorrGan) shown here.

https://gmarti.gitlab.io/qfin/2019/12/04/hierarchical-risk-parity-part-3.html

**Reply by dcajasn** [reply]
Hi @iron-marshal,

About that, there are several ways to do that. I prefer use a rolling window approach using paths of returns from historical sample, montecarlo, boostrapping or other return generation technique. However, I don't plan to add this feature because Riskfolio-Lib is just a portfolio optimization library. There are several portfolio backtesting libraries that you can use to check the out of sample performance of each model.

Best,
Dany

**Comment by jdavidk** [comment]
@dcajasn --

First, what a great library -- thank you so much for taking the time to pull together all of these research papers and convert them into working Python code!

Have you looked into incorporating the 'Maximum Decorrelation Portfolio' (Christoffersen et al. 2010 )into the 'optimization' function? From my reading (and similar to the Maximum Diversification Portfolio 'trick' that you laid out above), it might be as simple as a 'MinRisk' optimization where the covariance matrix is simply replaced with the correlation matrix...but I'm not certain about that.

Regardless, I appreciate you considering this...and again, thank you for all of your efforts!

**Reply by dcajasn** [reply]
Hi @jdavidk,

This kind of special models can be solved using custom mean vector and covariance matrix, also there is an example in docs to do that. For this reason, I don't plan to add this feature to riskfolio-lib.

Best,
Dany

**Comment by athanszheyin** [comment]
Hi @dcajasn great work on the library! 
Currently in the Worst Case Mean Variance Portfolio Optimization section, the uncertainty set for return/risk estimation is generated via bootstrapping. I am wondering if we could add an alternative to compute the uncertainty set from a structured model (say, factor model) instead of simulation? 
I found a series of literature on this approach by Professor D. GOLDFARB and G. IYENGAR. The most cited one being [here](https://www.researchgate.net/profile/Garud-Iyengar/publication/220442829_Robust_Portfolio_Selection_Problems/links/00b7d53b556319aded000000/Robust-Portfolio-Selection-Problems.pdf).

Best regards and thanks again for all the efforts!

**Reply by dcajasn** [reply]
Hi @athanszheyin, 

Thanks for your support, about your question. When I was investigating worst case models most papers talk about how to build the uncertainty sets but they don't talk about how to find the parameters of the uncertainty sets, so after investigating I decide to include three options:
- Bootstrapping: this method generates uncertainty sets that reflects the distribution of the sample without any assumption.
- Assume normality: in this case I use a multivariate normal distribution for the mean, and a Wishart distribution for the covariance.
- Delta: for box constraints assume an increase of delta percent above and below the sample parameter.

About the paper, I read that paper sometime ago, but I decided to not implement that model in that moment because it is too complex and few people will use it. I think that in medium term I will add it to Riskfolio-Lib. On the other hand, I you like worst case portfolio optimization, you can read the book of Feng and Palomar [A Signal Processing Perspective on Financial Engineering](https://www.danielppalomar.com/publications/).

Best,
Dany

**Reply by athanszheyin** [reply]
Hi @dcajasn ,

Thanks a lot for the prompt response! The uncertainty set is definitely the key to worst case robust optimization, and I share the pain with you that most literature were vague on the parameter settings of the uncertainty sets. This recent paper actually provides some practical guidance on implementing the uncertainty set: https://www.researchgate.net/profile/Romain-Perchet/publication/337854609_A_Practical_Guide_to_Robust_Portfolio_Optimization/links/5f1020e592851c1eff156ad5/A-Practical-Guide-to-Robust-Portfolio-Optimization.pdf

The book/slides by Feng and Palomar and the book by Kolm and Fabozzi are canonical in the field of robust portfolio optimization, only that both were a bit outdated. It seems the quantitative finance academia has alleviated their passion on robust optimization in recent years.

Thanks for the explanation anyhow! And I always enjoy such sort of discussion!

**Reply by dcajasn** [reply]
Thanks for the paper, I will read it asap. On the other hand, it's true that quantitative finance academia press less attention to worst case optimization, I think that is because worst case optimization still suffer of the same problems of traditional mean variance optimization: corner solutions and lack of robustness of solutions.

**Reply by athanszheyin** [reply]
Hi @dcajasn , 

I was playing with the worst case mean variance optimization code, I noticed in **line 934 and 935 of Portfolio.py**
`mu_u = mu + st.norm.ppf(1 - q / 2) * np.diag(cov) / n**2`
`mu_l = mu - st.norm.ppf(1 - q / 2) * np.diag(cov) / n**2`

My question is that why are we using second moment`cov/n**2` to represent the box uncertainty diameter to the first moment `mu`? To be comparable with the classic confidence interval definition, should we use some sort of `(cov/n)**0.5` here?

**Reply by dcajasn** [reply]
Hi @athanszheyin,

You are right, it is my mistake. I also found a mistake in line 942  `cov_l = np.percentile(A, q=q, axis=0)` must be `cov_l = np.percentile(A, q=q / 2, axis=0)`.

Best,
Dany

**Comment by cheukhin1024** [comment]
> Hi people, I would like to ask you what new features, portfolio optimization models or techniques would you like to add to Riskfolio-Lib. I think that I cover all available convex portfolio optimization models but maybe you know more models.

I hope Riskfolib can support Apache Spark and Koalas. I tried to replace Pandas with Spark, but it doesn't work.

**Reply by dcajasn** [reply]
Hi @cheukhin1024,

By the moment I don't have plans to add support for apache spark or similar libraries.

Best
Dany

**Comment by cpowr** [comment]
Hi @dcajasn,
Your library is solid. Congratulations on incorporating some of the best models in literature.
I am curious as to whether you are thinking about adding transaction costs to the hierarchical clustering methods. Not sure if I missed it if you already have. 
Thanks!
